File: app/src/main/java/ohi/andre/consolelauncher/commands/tuixt/TuixtDialog.java
package ohi.andre.consolelauncher.commands.tuixt;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Handler;
import android.os.Looper;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import java.util.List;

import ohi.andre.consolelauncher.tuils.LauncherSystemUi;

public final class TuixtDialog {

    public interface ItemAction {
        void onItemSelected(int index);
    }

    public interface InputAction {
        void onInput(String value);
    }

    public interface ConfirmAction {
        void onConfirm();
    }

    public interface ContentFactory {
        View create(Dialog dialog);
    }

    private TuixtDialog() {
    }

    public static void showOptions(Context context, String title, List<String> items, ItemAction action) {
        new Handler(Looper.getMainLooper()).post(() -> {
            Dialog dialog = createDialog(context);
            LinearLayout content = createContent(context);

            for (int i = 0; i < items.size(); i++) {
                int index = i;
                TextView row = new TextView(context);
                row.setText(items.get(i).toUpperCase());
                TuixtTheme.styleListItem(context, row, false);
                LinearLayout.LayoutParams rowParams = new LinearLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT);
                rowParams.bottomMargin = TuixtTheme.dp(context, 8);
                content.addView(row, rowParams);
                row.setOnClickListener(v -> {
                    dialog.dismiss();
                    action.onItemSelected(index);
                });
            }

            dialog.setContentView(wrap(context, title, content, null));
            show(dialog);
        });
    }

    public static void showInput(Context context, String title, String hint, String positive, String negative, InputAction action) {
        new Handler(Looper.getMainLooper()).post(() -> {
            Dialog dialog = createDialog(context);
            LinearLayout content = createContent(context);

            EditText input = new EditText(context);
            input.setHint(hint);
            TuixtTheme.styleInput(context, input);
            content.addView(input, new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT));

            LinearLayout buttons = buttons(context, dialog, positive, negative, () -> action.onInput(input.getText().toString()));
            dialog.setContentView(wrap(context, title, content, buttons));
            show(dialog);
        });
    }

    public static void showConfirm(Context context, String title, String message, String positive, String negative, ConfirmAction action) {
        new Handler(Looper.getMainLooper()).post(() -> {
            Dialog dialog = createDialog(context);
            LinearLayout content = createContent(context);

            TextView messageView = new TextView(context);
            messageView.setText(message);
            messageView.setTextColor(TuixtTheme.textColor());
            messageView.setTypeface(ohi.andre.consolelauncher.tuils.Tuils.getTypeface(context));
            messageView.setTextSize(14);
            messageView.setPadding(
                    TuixtTheme.dp(context, 10),
                    TuixtTheme.dp(context, 12),
                    TuixtTheme.dp(context, 10),
                    TuixtTheme.dp(context, 12));
            content.addView(messageView, new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT));

            LinearLayout buttons = buttons(context, dialog, positive, negative, action);
            dialog.setContentView(wrap(context, title, content, buttons));
            show(dialog);
        });
    }

    public static void showContent(Context context, String title, View customContent, String positive, String negative, ConfirmAction action) {
        new Handler(Looper.getMainLooper()).post(() -> {
            Dialog dialog = createDialog(context);
            LinearLayout content = createContent(context);
            content.addView(customContent, new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT));

            LinearLayout buttons = buttons(context, dialog, positive, negative, action);
            dialog.setContentView(wrap(context, title, content, buttons));
            show(dialog);
        });
    }

    public static void showCustom(Context context, String title, ContentFactory factory) {
        new Handler(Looper.getMainLooper()).post(() -> {
            Dialog dialog = createDialog(context);
            View content = factory.create(dialog);
            dialog.setContentView(wrap(context, title, content, null));
            show(dialog);
        });
    }

    private static Dialog createDialog(Context context) {
        Dialog dialog = new Dialog(context);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        return dialog;
    }

    private static LinearLayout createContent(Context context) {
        LinearLayout content = new LinearLayout(context);
        content.setOrientation(LinearLayout.VERTICAL);
        content.setPadding(0, 0, 0, 0);
        return content;
    }

    private static View wrap(Context context, String title, View content, View buttons) {
        FrameLayout root = new FrameLayout(context);
        root.setPadding(
                TuixtTheme.dp(context, 18),
                TuixtTheme.dp(context, 18),
                TuixtTheme.dp(context, 18),
                TuixtTheme.dp(context, 18));

        FrameLayout container = new FrameLayout(context);

        LinearLayout panel = new LinearLayout(context);
        panel.setOrientation(LinearLayout.VERTICAL);
        panel.setPadding(
                TuixtTheme.dp(context, 14),
                TuixtTheme.dp(context, 36),
                TuixtTheme.dp(context, 14),
                TuixtTheme.dp(context, 14));
        TuixtTheme.stylePanel(context, panel);

        ScrollView scrollView = new ScrollView(context);
        scrollView.setFillViewport(false);
        scrollView.addView(content, new ScrollView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT));
        panel.addView(scrollView, new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT));

        if (buttons != null) {
            panel.addView(buttons);
        }

        FrameLayout.LayoutParams panelParams = new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT);
        panelParams.topMargin = TuixtTheme.dp(context, 12);
        container.addView(panel, panelParams);

        TextView header = new TextView(context);
        header.setText(title.toUpperCase());
        TuixtTheme.styleHeader(context, header);
        FrameLayout.LayoutParams headerParams = new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT);
        headerParams.gravity = Gravity.TOP | Gravity.START;
        headerParams.leftMargin = TuixtTheme.dp(context, 42);
        container.addView(header, headerParams);

        FrameLayout contentHost = TuixtLayout.addFoldAwareHost(context, root, ViewGroup.LayoutParams.WRAP_CONTENT);
        contentHost.addView(container, new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                Gravity.CENTER));

        return root;
    }

    private static LinearLayout buttons(Context context, Dialog dialog, String positive, String negative, ConfirmAction action) {
        LinearLayout row = new LinearLayout(context);
        row.setOrientation(LinearLayout.HORIZONTAL);
        row.setGravity(Gravity.CENTER_VERTICAL);
        row.setPadding(0, TuixtTheme.dp(context, 12), 0, 0);

        TextView cancel = new TextView(context);
        cancel.setText(negative.toUpperCase());
        TuixtTheme.styleButton(context, cancel, false);
        cancel.setOnClickListener(v -> dialog.dismiss());
        row.addView(cancel, new LinearLayout.LayoutParams(
                0,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                1));

        View spacer = new View(context);
        row.addView(spacer, new LinearLayout.LayoutParams(TuixtTheme.dp(context, 12), 1));

        TextView confirm = new TextView(context);
        confirm.setText(positive.toUpperCase());
        TuixtTheme.styleButton(context, confirm, true);
        confirm.setOnClickListener(v -> {
            dialog.dismiss();
            action.onConfirm();
        });
        row.addView(confirm, new LinearLayout.LayoutParams(
                0,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                1));

        return row;
    }

    private static void show(Dialog dialog) {
        dialog.show();
        LauncherSystemUi.applyFullscreen(dialog);
        Window window = dialog.getWindow();
        if (window != null) {
            window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            window.setDimAmount(0f);
            WindowManager.LayoutParams params = new WindowManager.LayoutParams();
            params.copyFrom(window.getAttributes());
            params.width = WindowManager.LayoutParams.MATCH_PARENT;
            params.height = WindowManager.LayoutParams.WRAP_CONTENT;
            window.setAttributes(params);
        }
    }
}

File: app/src/main/java/ohi/andre/consolelauncher/commands/tuixt/TuixtDialog.kt
package ohi.andre.consolelauncher.commands.tuixt

import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import ohi.andre.consolelauncher.commands.tuixt.TuixtLayout.addFoldAwareHost
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.dp
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.styleButton
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.styleHeader
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.styleInput
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.styleListItem
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.stylePanel
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.textColor
import ohi.andre.consolelauncher.tuils.LauncherSystemUi.applyFullscreen
import ohi.andre.consolelauncher.tuils.Tuils
import java.util.Locale
import ohi.andre.consolelauncher.tuils.LauncherSystemUi

object TuixtDialog {
    fun showOptions(
        context: Context,
        title: String,
        items: List<String?>,
        action: ItemAction
    ) {
        Handler(Looper.getMainLooper()).post(Runnable {
            val dialog = createDialog(context)
            val content = createContent(context)

            for (i in items.indices) {
                val index = i
                val row = TextView(context)
                row.setText(items.get(i)!!.uppercase(Locale.getDefault()))
                styleListItem(context, row, false)
                val rowParams = LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
                rowParams.bottomMargin = dp(context, 8f)
                content.addView(row, rowParams)
                row.setOnClickListener(View.OnClickListener { v: View? ->
                    dialog.dismiss()
                    action.onItemSelected(index)
                })
            }

            dialog.setContentView(wrap(context, title, content, null))
            show(dialog)
        })
    }

    fun showInput(
        context: Context,
        title: String,
        hint: String?,
        positive: String,
        negative: String,
        action: InputAction
    ) {
        Handler(Looper.getMainLooper()).post(Runnable {
            val dialog = createDialog(context)
            val content = createContent(context)

            val input = EditText(context)
            input.setHint(hint)
            styleInput(context, input)
            content.addView(
                input, LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
            )

            val buttons = buttons(
                context,
                dialog,
                positive,
                negative,
                ConfirmAction { action.onInput(input.getText().toString()) })
            dialog.setContentView(wrap(context, title, content, buttons))
            show(dialog)
        })
    }

    fun showConfirm(
        context: Context,
        title: String,
        message: String?,
        positive: String,
        negative: String,
        action: ConfirmAction
    ) {
        Handler(Looper.getMainLooper()).post(Runnable {
            val dialog = createDialog(context)
            val content = createContent(context)

            val messageView = TextView(context)
            messageView.setText(message)
            messageView.setTextColor(textColor())
            messageView.setTypeface(Tuils.getTypeface(context))
            messageView.setTextSize(14f)
            messageView.setPadding(
                dp(context, 10f),
                dp(context, 12f),
                dp(context, 10f),
                dp(context, 12f)
            )
            content.addView(
                messageView, LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
            )

            val buttons = buttons(context, dialog, positive, negative, action)
            dialog.setContentView(wrap(context, title, content, buttons))
            show(dialog)
        })
    }

    fun showContent(
        context: Context,
        title: String,
        customContent: View?,
        positive: String,
        negative: String,
        action: ConfirmAction
    ) {
        Handler(Looper.getMainLooper()).post(Runnable {
            val dialog = createDialog(context)
            val content = createContent(context)
            content.addView(
                customContent, LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
            )

            val buttons = buttons(context, dialog, positive, negative, action)
            dialog.setContentView(wrap(context, title, content, buttons))
            show(dialog)
        })
    }

    fun showCustom(context: Context, title: String, factory: ContentFactory) {
        Handler(Looper.getMainLooper()).post(Runnable {
            val dialog = createDialog(context)
            val content = factory.create(dialog)
            dialog.setContentView(wrap(context, title, content, null))
            show(dialog)
        })
    }

    private fun createDialog(context: Context): Dialog {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        return dialog
    }

    private fun createContent(context: Context?): LinearLayout {
        val content = LinearLayout(context)
        content.setOrientation(LinearLayout.VERTICAL)
        content.setPadding(0, 0, 0, 0)
        return content
    }

    private fun wrap(context: Context, title: String, content: View?, buttons: View?): View {
        val root = FrameLayout(context)
        root.setPadding(
            dp(context, 18f),
            dp(context, 18f),
            dp(context, 18f),
            dp(context, 18f)
        )

        val container = FrameLayout(context)

        val panel = LinearLayout(context)
        panel.setOrientation(LinearLayout.VERTICAL)
        panel.setPadding(
            dp(context, 14f),
            dp(context, 36f),
            dp(context, 14f),
            dp(context, 14f)
        )
        stylePanel(context, panel)

        val scrollView = ScrollView(context)
        scrollView.setFillViewport(false)
        scrollView.addView(
            content, FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        )
        panel.addView(
            scrollView, LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        )

        if (buttons != null) {
            panel.addView(buttons)
        }

        val panelParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        panelParams.topMargin = dp(context, 12f)
        container.addView(panel, panelParams)

        val header = TextView(context)
        header.setText(title.uppercase(Locale.getDefault()))
        styleHeader(context, header)
        val headerParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        headerParams.gravity = Gravity.TOP or Gravity.START
        headerParams.leftMargin = dp(context, 42f)
        container.addView(header, headerParams)

        val contentHost = addFoldAwareHost(context, root, ViewGroup.LayoutParams.WRAP_CONTENT)
        contentHost.addView(
            container, FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                Gravity.CENTER
            )
        )

        return root
    }

    private fun buttons(
        context: Context,
        dialog: Dialog,
        positive: String,
        negative: String,
        action: ConfirmAction
    ): LinearLayout {
        val row = LinearLayout(context)
        row.setOrientation(LinearLayout.HORIZONTAL)
        row.setGravity(Gravity.CENTER_VERTICAL)
        row.setPadding(0, dp(context, 12f), 0, 0)

        val cancel = TextView(context)
        cancel.setText(negative.uppercase(Locale.getDefault()))
        styleButton(context, cancel, false)
        cancel.setOnClickListener(View.OnClickListener { v: View? -> dialog.dismiss() })
        row.addView(
            cancel, LinearLayout.LayoutParams(
                0,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                1f
            )
        )

        val spacer = View(context)
        row.addView(spacer, LinearLayout.LayoutParams(dp(context, 12f), 1))

        val confirm = TextView(context)
        confirm.setText(positive.uppercase(Locale.getDefault()))
        styleButton(context, confirm, true)
        confirm.setOnClickListener(View.OnClickListener { v: View? ->
            dialog.dismiss()
            action.onConfirm()
        })
        row.addView(
            confirm, LinearLayout.LayoutParams(
                0,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                1f
            )
        )

        return row
    }

    private fun show(dialog: Dialog) {
        dialog.show()
        applyFullscreen(dialog)
        val window = dialog.getWindow()
        if (window != null) {
            window.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            window.setDimAmount(0f)
            val params = WindowManager.LayoutParams()
            params.copyFrom(window.getAttributes())
            params.width = WindowManager.LayoutParams.MATCH_PARENT
            params.height = WindowManager.LayoutParams.WRAP_CONTENT
            window.setAttributes(params)
        }
    }

    fun interface ItemAction {
        fun onItemSelected(index: Int)
    }

    fun interface InputAction {
        fun onInput(value: String?)
    }

    fun interface ConfirmAction {
        fun onConfirm()
    }

    fun interface ContentFactory {
        fun create(dialog: Dialog?): View?
    }
}
